src/main/java/frc/robot/subsystems/Drivetrain.java

```diff
@@ -15,31 +15,83 @@
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
 import frc.robot.Constants;
 import frc.robot.drive.MotorSpeedController;
+import frc.robot.drive.SparkMaxController;
+import frc.robot.drive.SpeedControllerEncoder;
+import frc.robot.drive.TalonController;
 
 public class Drivetrain extends SubsystemBase {
     SpeedControllerGroup leftMotorGroup;
-    MotorSpeedController leftMotorLeader;
-    MotorSpeedController leftMotorFollower = null;
+    SpeedControllerEncoder leftMotorLeader;
+    SpeedControllerEncoder leftMotorFollower = null;
 
     SpeedControllerGroup rightMotorGroup;
-    MotorSpeedController rightMotorLeader;
-    MotorSpeedController rightMotorFollower = null;
+    SpeedControllerEncoder rightMotorLeader;
+    SpeedControllerEncoder rightMotorFollower = null;
     
     DifferentialDrive diffDrive;
 
     public Drivetrain() {
         super();
 
-        leftMotorLeader = new MotorSpeedController(Constants.DRIVE_MOTOR_LEFT_LEADER_ID, Constants.DRIVE_MOTOR_LEFT_LEADER_TYPE);
-        rightMotorLeader = new MotorSpeedController(Constants.DRIVE_MOTOR_RIGHT_LEADER_ID, Constants.DRIVE_MOTOR_RIGHT_LEADER_TYPE);
+        switch (Constants.DRIVE_MOTOR_LEFT_LEADER_TYPE) {
+            case TALON_SRX:
+                leftMotorLeader = new TalonController(Constants.DRIVE_MOTOR_LEFT_LEADER_ID);
+                break;
+
+            case SPARK_MAX_BRUSHED:
+                leftMotorLeader = new SparkMaxController(Constants.DRIVE_MOTOR_LEFT_LEADER_ID, MotorType.kBrushed);
+                break;
+
+            case SPARK_MAX_BRUSHLESS:
+                leftMotorLeader = new SparkMaxController(Constants.DRIVE_MOTOR_LEFT_LEADER_ID, MotorType.kBrushless);
+                break;
+        }
```

Comment:
I wonder if it would be worth creating a `MotorControllerFactory` class to do what these switch statements do, a class with a static function something like `MotorControllerFactory::create(motorType, motorID)`. This would be useful everywhere we need a motor, not just drive. It also makes it easier to add new motor types in the future.